common/src/main/java/io/homeassistant/companion/android/common/data/keychain/KeyChainRepositoryImpl.kt

```diff
@@ -0,0 +1,62 @@
+package io.homeassistant.companion.android.common.data.keychain
+
+import android.content.Context
+import android.security.KeyChain
+import io.homeassistant.companion.android.common.data.prefs.PrefsRepository
+import java.security.PrivateKey
+import java.security.cert.X509Certificate
+import java.util.concurrent.Executors
+import javax.inject.Inject
+
+class KeyChainRepositoryImpl@Inject constructor() : KeyChainRepository {
+
+    @Inject
+    lateinit var prefsRepository: PrefsRepository
+
+    private var alias: String? = null
+    private var key: PrivateKey? = null
+    private var chain: Array<X509Certificate>? = null
+    private var isLoading: Boolean = false
+
+    override suspend fun getAlias(): String? {
+        return alias
+    }
+
+    override suspend fun load(context: Context, alias: String): Boolean {
+        if (alias == null) return isLoading
+
+        this.alias = alias
+        prefsRepository.saveKeyAlias(alias)
+
+        return load(context)
+    }
+
+    override suspend fun load(context: Context): Boolean {
+        if (alias == null) {
+            alias = prefsRepository.getKeyAlias()
+        }
+
+        if (alias != null && !isLoading && (key == null || chain == null)) {
+            isLoading = true // TODO: need proper sync
+            Executors.newSingleThreadExecutor().execute {
+                if (chain == null) {
+                    chain = KeyChain.getCertificateChain(context, alias!!)
+                }
+                if (key == null) {
+                    key = KeyChain.getPrivateKey(context, alias!!)
+                }
+                isLoading = false
+            }
+        }
```

Comment:
Why are you spawning a thread for this?  Why can't we just use a coroutine?

Reply:
It was to be sure this code will not run out of the main thread. Do we have the guarantee the corouting won't be running out of the main thread or can we somehow force this?

Reply:
I just realised I wrote a comment about this but it is not well located. In essence, calling `getCertificateChain` or `getPrivateKey` will block until a connection is received, if called from the main looper then there won't be any more dispatch and the app will hang. It is documented [here](https://developer.android.com/reference/android/security/KeyChain).
So it is important, but maybe there is a cleaner way to this (I am not familiar at all with the Android SDK and the Kotlin API)

Reply:
I suggest wrapping the whole `load(Context)` method's body into a `withContext(Dispatchers.IO) { ... }`, like this:

```
override suspend fun load(context: Context): Boolean = withContext(Dispatchers.IO) {
```

And you can remove the `Executor` altogether, and make the logic sequential. I don't know if the 5 sec timeout is important, but if it is, I can help to come up with a coroutine-based solution for that.

Reply:
Thank you this worked well and the 5 sec timeout was purely arbitrary!